src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs

```diff
@@ -12688,5 +12942,640 @@ public void RefStructHandler_DynamicInHole(string expression)
+        {
+            var code = @"
+using System;
+using System.Runtime.CompilerServices;
+
+C.M(1, $""Literal"");
+
+class C
+{
+    public static void M(int i, [InterpolatedStringHandlerArgument(""i"")] CustomHandler c) => Console.WriteLine(c.ToString());
+}
+
+[InterpolatedStringHandler]
+partial struct CustomHandler
+{
+    private string _s = null;
+    public CustomHandler(int literalLength, int formattedCount, int i1, out bool isValid, int i2 = 2) { _s = i1.ToString() + i2.ToString(); isValid = false; }
+    public void AppendLiteral(string s) => _s += s;
+    public override string ToString() => _s;
+}
+";
+
+            var verifier = CompileAndVerify(new[] { code, InterpolatedStringHandlerArgumentAttribute, InterpolatedStringHandlerAttribute }, expectedOutput: @"12");
+            verifier.VerifyDiagnostics();
+
+            verifier.VerifyIL("<top-level-statements-entry-point>", @"
+{
+  // Code size       42 (0x2a)
+  .maxstack  6
+  .locals init (int V_0,
+                CustomHandler V_1,
+                bool V_2)
+  IL_0000:  ldc.i4.1
+  IL_0001:  stloc.0
+  IL_0002:  ldloc.0
+  IL_0003:  ldc.i4.7
+  IL_0004:  ldc.i4.0
+  IL_0005:  ldloc.0
+  IL_0006:  ldloca.s   V_2
+  IL_0008:  ldc.i4.2
+  IL_0009:  newobj     ""CustomHandler..ctor(int, int, int, out bool, int)""
+  IL_000e:  stloc.1
+  IL_000f:  ldloc.2
+  IL_0010:  brfalse.s  IL_0021
+  IL_0012:  ldloca.s   V_1
+  IL_0014:  ldstr      ""Literal""
+  IL_0019:  call       ""void CustomHandler.AppendLiteral(string)""
+  IL_001e:  ldc.i4.1
+  IL_001f:  br.s       IL_0022
+  IL_0021:  ldc.i4.0
+  IL_0022:  pop
+  IL_0023:  ldloc.1
+  IL_0024:  call       ""void C.M(int, CustomHandler)""
+  IL_0029:  ret
+}
+");
+        }
+
+        [Fact]
+        public void HandlerExtensionMethod_01()
+        {
+            var code = @"
+$""Test"".M();
+
+public static class StringExt
+{
+    public static void M(this CustomHandler handler) => throw null;
+}
+";
+
+            var comp = CreateCompilation(new[] { code, GetInterpolatedStringCustomHandlerType("CustomHandler", "struct", useBoolReturns: false) });
+            comp.VerifyDiagnostics(
+                // (2,1): error CS1929: 'string' does not contain a definition for 'M' and the best extension method overload 'StringExt.M(CustomHandler)' requires a receiver of type 'CustomHandler'
+                // $"Test".M();
+                Diagnostic(ErrorCode.ERR_BadInstanceArgType, @"$""Test""").WithArguments("string", "M", "StringExt.M(CustomHandler)", "CustomHandler").WithLocation(2, 1)
+            );
+        }
+
+        [Fact]
+        public void HandlerExtensionMethod_02()
+        {
+            var code = @"
+using System.Runtime.CompilerServices;
+
+var s = new S1();
+s.M($"""");
+
+public struct S1
+{
+    public int Field = 1;
+}
+
+public static class S1Ext
+{
+    public static void M(this S1 s, [InterpolatedStringHandlerArgument("""")] CustomHandler c) => throw null;
```

Comment:
Is this covered by the spec?

Reply:
I'm not sure what you're getting at here. There's nothing special about extension methods: they don't have receivers. C# just allows for some fancy sugar at the call site. There is no `this`.